src/Symfony/Component/EventDispatcher/InvokableEventSubscriberTrait.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\EventDispatcher;
+
+use Symfony\Contracts\EventDispatcher\Event as ContractsEvent;
+
+/**
+ * Helper trait for single-event subscription.
+ *
+ * @author Yonel Ceruto <yonelceruto@gmail.com>
+ */
+trait InvokableEventSubscriberTrait
+{
+    /**
+     * {@inheritdoc}
+     */
+    final public static function getSubscribedEvents(): array
+    {
+        return [static::getEventName() => ['__invoke', static::getPriority()]];
+    }
+
+    protected static function getEventName(): string
+    {
+        $method = new \ReflectionMethod(static::class, '__invoke');
+
+        if (0 === $method->getNumberOfParameters()) {
+            throw new \LogicException(sprintf('The Event name cannot be inferred from "%s::%s()" method without parameters, you must to define one or implement the "%s::getEventName()" method to return it explicitly.', static::class, $method->getName(), static::class));
+        }
+
+        $parameter = $method->getParameters()[0];
+
+        if (null === $type = $parameter->getType()) {
+            throw new \LogicException(sprintf('The Event name cannot be inferred from "%s::%s()" method without type-hint the "$%s" parameter, you must to type-hint it or implement the "%s::getEventName()" method to return it explicitly.', ContractsEvent::class, $method->getName(), $parameter->getName(), static::class));
+        }
+
+        if ($type->isBuiltin()) {
+            throw new \LogicException(sprintf('The "$%s" parameter defined in "%s::%s()" method must be a class of "%s", "%s" given.', $parameter->getName(), static::class, $method->getName(), ContractsEvent::class, $type->getName()));
+        }
+
+        return $type->getName();
+    }
+
+    protected static function getPriority(): int
```

Comment:
Playing with this last version I think that not making the `getEventName()` and `getPriority()` methods discoverable will defect the another goal: provides configuration hooks.

We have also a lot of good examples of this kind of traits in core: `MicroKernelTrait`, `ResponseTrait`, `ContainerAwareTrait`, etc, etc.

Reply:
Anyway making it `private` would not prevent you from overriding the method, KISS ;)

Reply:
I don't understand, what's the relationship with discoverability? To me the method should be private and getSubscribedEvents should not be final. Open/closed principle.

Reply:
Open/closed principle will affect only the visibility for IDEs, you can to override the private methods anyway, no?

Reply:
Making `getSubscribedEvents` not final, `getEventName()` and `getPriority()` private, this is what PHPStorm will show:
![visibility-nicolas-2](https://user-images.githubusercontent.com/2028198/64478791-b2a74500-d17b-11e9-835e-ec52cb98717f.png)
![visibility-nicolas](https://user-images.githubusercontent.com/2028198/64478784-87bcf100-d17b-11e9-9be0-e0f039568e25.png)

Reply:
this is what currently shows:
![visibility-yonel-2](https://user-images.githubusercontent.com/2028198/64478801-d4083100-d17b-11e9-894d-c10a58bbe355.png)
![visibility-after](https://user-images.githubusercontent.com/2028198/64478802-d8cce500-d17b-11e9-983f-d888e5a17924.png)

Reply:
Better DX, don't you think?

Reply:
DX apart, you are totally right!!

Reply:
That's a bug in phpstorm. When a method is imported from a trait, it should be listed as any other for override. final or private is totally fine overriding from a trait.

phpstorm-driven-development is not a thing, and should not be one!

I'm pretty confident about my suggestion above.

Reply:
Fair enough :+1: